### ISSUE TITLE
Bundle public keys when signing AppImages

### ISSUE THREAD

**Issue by TheAssassin** [AppImage/AppImageKit#696]
Follow-up for https://github.com/AppImage/AppImageUpdate/issues/16#issuecomment-370924588.

We need to bundle public keys of the signatures in order to be able to verify the signatures of AppImages.

The keys probably need to be bundled in the same way the signatures are bundled, and therefore require our runtime to add another maybe ~3-5kiB ELF header field.

**Comment by probonopd** [comment]
Isn't the point of the whole thing that the user somehow is expected to add the keys he trusts to the keyring (or have the distro ship with some "trusted" keys)?

**Comment by TheAssassin** [comment]
No... https://github.com/AppImage/AppImageUpdate/issues/16#issuecomment-370928960. You need the key file to verify whether the signature applies at all. Then, you can check whether the key can be trusted. That latter part is discussed in https://github.com/AppImage/AppImageUpdate/issues/16#issuecomment-368320196.

**Comment by probonopd** [comment]
Does it make sense to check whether the signature applies if the signature is not trusted? (Possibly it does... but how many users will actually know the difference?)

**Comment by TheAssassin** [comment]
@probonopd well of course you don't necessarily have to check whether the signature applies before comparing the key IDs. It makes sense to check the validity of a signature _beforehand_ however. A bad signature is the way more critical issue. Also, by checking it beforehand, I need to call `gpg2` just once.
Anyway, you need to have the key in order to be able to verify signatures at all, because even if you had decided to trust the key, without the key file, you couldn't validate the signatures. So, we have to bundle them anyway.

**Comment by probonopd** [comment]
Err... aren't trusted keys in your keychain? We could only attempt to check signatures if we find a trusted key in the keychain for that signature. Because an untrusted signature doesn't mean anything useful to the user anyway.

Or am I totally wrong here?

**Comment by TheAssassin** [comment]
@probonopd please decide on _one_ issue where to ask all these (pardon) "noob" questions...

You _are_ totally wrong here. We agreed on a trust model. And that is discussed in https://github.com/AppImage/AppImageUpdate/issues/16.

TL;DR: I trust the old AppImage. Therefore I trust the key it was signed with. Therefore I can use that key to verify the new AppImage. But I need the keyfile. And I can only get it if it's on a keyserver (unlike) or it has been bundled.

**Comment by probonopd** [comment]
> TL;DR: I trust the old AppImage. Therefore I trust the key it was signed with. Therefore I can use that key to verify the new AppImage. But I need the keyfile. And I can only get it if it's on a keyserver (unlike) or it has been bundled.

Makes sense. So +1 for bundling the key, to be used in the scenario you are describing.
